staging/src/k8s.io/apiserver/pkg/storage/cacher/cacher.go

```diff
@@ -592,17 +592,22 @@ func (c *Cacher) Watch(ctx context.Context, key string, opts storage.ListOptions
 		identifier,
 	)
 
+	// note that c.waitUntilWatchCacheFreshAndForceAllEvents must be called without
+	// the c.watchCache.RLock held otherwise we are at risk of a deadlock
+	// mainly because c.watchCache.processEvent method won't be able to make progress
+	forceAllEvents, err := c.waitUntilWatchCacheFreshAndForceAllEvents(ctx, requestedWatchRV, opts)
```

Comment:
How did you find this? Can you please add a test for it (that was failing or at least flaking before)?

Reply:
Good idea, I modified `TestWaitUntilWatchCacheFreshAndForceAllEvents` test. 
Without this patch the test will timeout and fail. 
I think the default timeout is 10 min.
I think that's okay but if you want we could provide a custom timeout for this test.